File: 2022/03/sacks-of-joy.py
with open('input.txt', 'r') as f:
    a = [i.strip() for i in f.readlines()]

def get_priority(item_list: list) -> int:
    return sum([(ord(i) - 96) if ord(i) >= 97 else (ord(i) - 38) for i in item_list])

# Part One
jungle_sacks_common = [''.join(set(i[:len(i)//2]).intersection(i[len(i)//2:])) for i in a]
print(get_priority(jungle_sacks_common))

# Part Two
groups_of_sacks = [''.join(set.intersection(*[set(a[j]) for j in range(i, i+3)])) for i in range(0, len(a), 3)]
print(get_priority(groups_of_sacks))

File: 2022/03/sacks-of-joy.jl
data = readlines("input.txt")

function priority(p::Set{Char})
    total = 0
    for i in p
        if occursin(r"[A-Z]", string(i))
            total += (Int(i) - 38)
        else
            total += (Int(i) - 96)
        end
    end
    return total
end

common_items = [Set.([d[1:length(d)÷2], d[length(d)÷2+1:end]]) for d in data]
total = sum(map(priority, [intersect(i[1], i[2]) for i in common_items]))
@show total

data = Set.(data)
badges = sum(map(priority, [intersect([data[d+k] for k in 0:2]...) for d in 1:3:length(data)]))
@show badges
